### PR TITLE
Dynamic onnx gemm bias

### PR DESCRIPTION
**Pull request by CharlieL7**
* Adds support for parsing dynamic ONNX gemm bias input `C`
* Require this functionality for onnx model zoo's version of Resnet50
* Not sure why it was checking for `C.elements() > 0` or how it was working for the `gemm_no_C` test
  * ONNX parses scalar to migraphx.shape with 1 element and stride of 0 
* Changed code to not add `beta` literal and multiply instruction if it is 1.0

### CODE REVIEW

**Comment by umangyadav** [comment]
> Not sure why it was checking for C.elements() < 0 or how it was working for the gemm_no_C test
 
 Where was this ? I see it is checking for  `element() > 0`. 

> Removed beta literal if it is 1 

This wasn't necessary. It would be cleaned up by simplify_algebra ( x * 1 = x) and dead-code-elimination.

**Comment by CharlieL7** [comment]
> > Not sure why it was checking for C.elements() < 0 or how it was working for the gemm_no_C test
> 
> Where was this ? I see it is checking for `element() > 0`.
> 
> > Removed beta literal if it is 1
> 
> This wasn't necessary. It would be cleaned up by simplify_algebra ( x * 1 = x) and dead-code-elimination.

* Typo, was C.elements > 0
* Agreed, alpha was not added this way so I changed beta to be consistent.

**Review by umangyadav — APPROVED** [review]

**Comment by umangyadav** [comment]
> Agreed, alpha was not added this way so I changed beta to be consistent. 

We could have removed code related to `alpha ==1` check.  That would have simplified code instead of adding similar checks for `beta`. but Fine for now.

**Comment by umangyadav** [comment]
> Not sure why it was checking for C.elements() > 0 or how it was working for the gemm_no_C test

It worked because of dead code elimination. Inside `onnx_test`, it is calling `optimize_onnx()`.

**Review by umangyadav — CHANGES_REQUESTED** [review]

File: src/onnx/parse_gemm.cpp [hunk 1]
```diff
@@ -90,41 +90,45 @@ struct parse_gemm : op_parser<parse_gemm>
...
         {
-            // TODO: support dynamic C input
-            if(std::any_of(args.cbegin(), args.cend(), [](auto in_arg) {
-                   return in_arg->get_shape().dynamic();
-               }))
+            if(not float_equal(beta, 0.0f))
             {
-                MIGRAPHX_THROW("PARSE_GEMM: C input not handled for dynamic input shapes");
-            }
-            if(not float_equal(beta, 0.0f) and args[2]->get_shape().elements() > 0)
```
[the remark this answers is not in the document]
Comment (CharlieL7) [reply src/onnx/parse_gemm.cpp:104]: `shape.elements()` only checks lens. I think the check only catches if we default construct a shape or similar, but doesn't look like the onnx parser does that.

[the remark this answers is not in the document]
Comment (TedThemistokleous) [reply src/onnx/parse_gemm.cpp:104]: I agree with @CharlieL7 on this. How often are we going to see that case either? That isn't valid if we have zero length + strides. Do you mean if we have zero len/strides but say the shape is still dynamic?

**Review by TedThemistokleous — COMMENTED** [review]
Looks good just the one comment we should resolve about adding back the elements() > 0 check.

File: src/onnx/parse_gemm.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (umangyadav) [reply src/onnx/parse_gemm.cpp:104]: I was thinking for scalar case. but `elements()` is not defined for dynamic shapes. So, I agree now that we shouldn't keep  `elements()>0` check. 

Saying that, how does the constructor makes a different ? We are taking shape of `migraphx::instruction` , and should be always be valid.

[the remark this answers is not in the document]
Comment (TedThemistokleous) [reply src/onnx/parse_gemm.cpp:104]: Some shape constructors will have the shape as empty with s{type, {}}  which comes up with empty items. That isn't valid onnx but its still possible with our shape constructor.

[the remark this answers is not in the document]
Comment (umangyadav) [reply src/onnx/parse_gemm.cpp:104]: But we are not taking shape of onnx's node. We are taking shape of migraphx::instruction.  Which can have empty lens. but even in  that case `elements()` is a valid call for static  shapes.

[the remark this answers is not in the document]
Comment (TedThemistokleous) [reply src/onnx/parse_gemm.cpp:104]: That's true. It would just resolve zero.

**Review by TedThemistokleous — APPROVED** [review]

**Review by umangyadav — APPROVED** [review]

**Review by pfultz2 — APPROVED** [review]

**Pull request merged**